generator/src/main/scala/grpcgateway/generators/SwaggerGenerator.scala

```diff
@@ -85,149 +83,167 @@ object SwaggerGenerator extends protocbridge.ProtocCodeGenerator with Descriptor
+          .add(s"- name: $prefix${field.getJsonName}")
+          .addIndented("in: query", s"type: ${t.name.toLowerCase}")
     }
   }
 
-  private def generateDefinition(d: Descriptor): PrinterEndo = { _
-    .add(d.getName + ":")
-    .indent
-    .add("type: object")
-    .add("properties: ")
-    .indent
-    .print(d.getFields.asScala) { case (printer, field) =>
-      if (field.isRepeated) {
-        printer
-          .add(field.getJsonName+":")
-          .indent
-          .add("type: array", "items:")
-          .indent
-          .call(generateDefinitionType(field))
-          .outdent
-          .outdent
-      } else {
-        printer
-          .add(field.getJsonName+":")
-          .indent
-          .call(generateDefinitionType(field))
-          .outdent
+  private def generateDefinition(d: Descriptor): PrinterEndo = {
+    _.add(d.getName + ":").indent
+      .add("type: object")
+      .add("properties: ")
+      .indent
+      .print(d.getFields.asScala) {
+        case (printer, field) =>
+          if (field.isRepeated) {
+            printer
+              .add(field.getJsonName + ":")
+              .indent
+              .add("type: array", "items:")
+              .indent
+              .call(generateDefinitionType(field))
+              .outdent
+              .outdent
+          } else {
+            printer
+              .add(field.getJsonName + ":")
+              .indent
+              .call(generateDefinitionType(field))
+              .outdent
+          }
       }
-    }
-    .outdent
-    .outdent
+      .outdent
+      .outdent
   }
 
   private def generateDefinitionType(field: FieldDescriptor): PrinterEndo = {
     field.getJavaType match {
       case JavaType.MESSAGE => _.add(s"""$$ref: "#/definitions/${field.getMessageType.getName}"""")
-      case JavaType.ENUM => _.add("type: string", "enum:").add(
-        field.getEnumType.getValues.asScala.map(v => s"- ${v.getName}"): _*
-      )
+      case JavaType.ENUM =>
+        _.add("type: string", "enum:").add(
+          field.getEnumType.getValues.asScala.map(v => s"- ${v.getName}"): _*
+        )
       case JavaType.INT  => _.add("type: integer", "format: int32")
       case JavaType.LONG => _.add("type: integer", "format: int64")
-      case t => _.add(s"type: ${t.name.toLowerCase}")
+      case t             => _.add(s"type: ${t.name.toLowerCase}")
     }
   }
 
   private def extractDefs(d: Descriptor): Set[Descriptor] = {
-    Set(d) ++ d.getFields.asScala.flatMap { f =>
-      f.getJavaType match {
-        case JavaType.MESSAGE => extractDefs(f.getMessageType)
-        case _ => Set.empty[Descriptor]
+    val explored: mutable.Set[Descriptor] = mutable.Set.empty
+    def extractDefsRec(d: Descriptor): Set[Descriptor] = {
+      if (explored.contains(d)) Set()
+      else {
+        explored.add(d)
+        Set(d) ++ d.getFields.asScala.flatMap { f =>
+          f.getJavaType match {
+            case _                => Set.empty[Descriptor]
```

Comment:
Exactly, I am talking about these 2 cases:
```scala
case _                                 => ...
case JavaType.MESSAGE => ...
```
The problem is that it now it always matches the first case and the 2nd one is always ignored.
If I switch them it might bring back the infinite recursion for cyclic reference in proto files.